document rulesets setting, support baseline files

braver opened the pull request:
- document rulesets setting
  the info is not available in the README,
  probably since we temporarily remove it years ago

Currently the linter does not support baseline files because we're using temp files. I want to look into that too.

braver commented:
Hi @kaste, I'm using PHP MD with baseline files (to suppress violations in existing code), and this is not currently supported.  See https://phpmd.org/documentation/index.html and scroll all the way down to the section about "Baseline". This does not work in conjunction with temp files, because the filename of the file that's being linted is matches against entries in that baseline file. And the utility does not support passing the real file's name via an argument. 

I don't think there is an easy way of making the plugin switch to real-files mode dynamically. Doing so is a bit outside my comfort level. Seemingly nobody runs into that problem, or nobody uses Sublime for PHP, that might also be true 😅 

So, unless you have any better ideas I think I might just change the behaviour to real files only?

kaste commented on linter.py:
```diff
@@ -4,9 +4,6 @@
-        # For now, do *NOT* capture 'filename' since phpmd reports 'real'
-        # paths, and Sublime and SL cannot map such paths to the original
-        # `view.file_name()`.
         r'(.+):(?P<line>\d+)\s*(?P<message>.+)$'
```
We do support `(?P<filename>)`

braver replied:
Oh, I missed that bit in the docs. It says...

> You can capture filename if the linter emits for multiple files at once. Very often linters don’t do that and then it’s best to omit capturing filename as SublimeLinter will figure out the correct value for you and you don’t run into problems with unnamed buffers or possibly symlinked files.

... so that still means capturing filenames here is useless, this linter reports one file at a time.

> Do you know how sublimelinter.com could point to the "latest" docs, it still refers the "stable" docs?

I have no idea how that stuff (Sphinx I think) works, and can't find anything in any obvious config files. 🤷🏻

kaste replied:
I thought this is a configuration of the domain sublimelinter.com thing.  Do you own this domain still btw?  Or is it public domain and in the org?

kaste replied:
And yeah we can remove the comment if you like, but the commit message is bonkers as you tell us `SL does not even support capturing filename...?` which is not really a standard commit message from your handbook.

braver replied:
> Do you own this domain still btw

No, one of the old maintainers has it, I need to look that up.

kaste replied:
That's also a configuration of readthedocs.org, "our" domain redirects to it.  But I don't think I have the credentials for it.

braver replied:
I think I found it.

<img width="845" alt="Screenshot 2023-08-04 at 14 19 57" src="https://github.com/SublimeLinter/SublimeLinter-phpmd/assets/2543659/fd4ed910-da92-4680-bc9e-6ad80f6d2b60">

I can also add your email address to that account and you should be able to access it too.

braver replied:
So "stable" should match the last release that's also available on package control.

kaste replied:
Yes stable is the release and latest is master.  But for SublimeLinter we don't release a new version if the docs change.  How did you login to readthedocs, with GitHub credentials?

braver replied:
my email address and a password

braver replied:
I don't seem to be able to add your email address to the account. I believe doing so, and sharing the password with you, would allow you to access the account as well. Maybe you also have another email address we could use?

<img width="508" alt="Screenshot 2023-08-04 at 17 24 24" src="https://github.com/SublimeLinter/SublimeLinter-phpmd/assets/2543659/41a5bbad-c408-4b65-949f-ecfa5bc272ed">

kaste replied:
You can't add me as a maintainer?  E.g. for mockito (which I maintain) I have 

![image](https://github.com/SublimeLinter/SublimeLinter-phpmd/assets/8558/c362bf5e-e69d-4c0c-9e64-d83fc7e27783)

braver replied:
Sent you an invite!

![Uploading Screenshot 2023-08-24 at 14.41.58.png…]()

kaste replied:
Got it!  Thanks.

kaste commented:
How do you want to use this?  Do you set `--baseline-file ...` in `args`?  Just switching to real files mode is not okay because  having a baseline file is probably more niche than having none.  (?)  We always prefer while-you-type linting.  

So a) we need to detect if the "baseline" mode is enabled.  For example we examine `args` or define a new setting and then read that.  b) if it is enabled, we dynamically switch to real file mode by setting   `self.tempfile_suffix = '-'` and flipping `$temp_file` to `$file_on_disk`.  (T.i. all done by implementing `def cmd(self):`.)

braver commented:
> How do you want to use this

When you first start using this you generate a baseline fine. If you're using a non-default filename or path for that you need to then specify that baseline file using command line arguments. However, if you use a rulesets file, it will look for a baseline file with the default name in the same directory as that rulesets file, and if it's there it will use it. So, I do pass the rulesets file as an argument, but not the baseline file: that's automatic.

That makes it a bit hard for us to detect, we'd basically have to do the same thing the utility is doing to figure out if there is a baseline file and if it should be used.

> having a baseline file is probably more niche than having none.

Perhaps. You can't really introduce phpmd into an existing codebase without one. But I have been running it just for the feedback, not as a requirement, and then it's fine without it. I assume more people use it like that. So yeah, it's perfectly useable without one, and while-you-type is not something you give up easily.

So, here's an idea. We could also read the lint mode setting. We use tempfiles, unless you select a different mode from "background". So that if you're using "load_save" (as I like to do for slow-ish linters like this one), you're signalling that it's ok if linting is not live, and we switch to real-files. We'd just need to describe that behaviour in the readme, and then implement it. That's easier, more flexible and more explicit, than trying to magically detect the right mode. And we don't need a new separate setting where it becomes weird that you can set it to both background linting and real-files at the same time (which won't work of course).

kaste commented:
Reading out the `lint_mode` sounds smart, clever guy, you.

kaste commented:
I'm basically fine with the approach *but* note that "manual" mode is ambiguous.  A user can switch to "manual" mode and still run in stdin or temp file mode on dirty, unsaved or unnamed views.  We only know that the buffer content and file-on-disk content match for "save" and "load" events.

Do you use the "manual" here bc the linter is too slow?

kaste commented on linter.py:
```diff
@@ -16,3 +12,10 @@ class Phpmd(Linter):
         'selector': 'embedding.php, source.php',
         '@rulesets:,': 'cleancode,codesize,controversial,design,naming,unusedcode'
     }
+
+    def cmd(self):
+        if self.settings['lint_mode'] != 'background':
+            self.tempfile_suffix = '-'
+            return ('phpmd', '${file_on_disk}', 'text')
+
+        return ('phpmd', '${temp_file}', 'text')
```
Of course this would not make the "manual" trigger safe to use.   That's a bit the question:  should we disallow a manual trigger because it would somehow result in garbage.

braver replied:
Indeed, I did not consider the state of the view. I was approaching as if I would set the required behaviour globally or in the project, and then start using the linter. What you describe would indeed be my concern, that a manual trigger would return different results from any other live mode. For instance, I sometimes just hit "lint this view" to make sure the linters have run, and that should be consistent with e.g. just saving the file etc.

I have a bit of a struggle getting back into the mindset of all the different modes and interactions SL supports .. it being Friday and all 😅 But my initial reaction would be that your proposal is overly subtle 🤔

kaste replied:
> I have a bit of a struggle getting back into the mindset of all the different modes and interactions SL supports

Well, I actually cheat and have to read the source. And the documentation about capturing filename I wrote yesterday.  

So your approach: the user sets "manual", hits "lint this view" on dirty buffers but it actually still lints the file on disk.  All this even when the users doesn't use any baseline stuff.  

The subtle approach:  the users uses baseline stuff, and for example "load_save" mode, and then uses "lint this view" on an unsaved file, then SL will switch to temporary filenames and phpmd will produce noise not respecting the baseline.  The user could now roll eyes and fix it by saving and linting again.  

The latter is slightly better as a) the user can roll-eyes and fix the situation and b) the other user not using baselines is not affected at all.  Am I correct?

All can probably be solved with an extra setting `only_lint_saved_files` and code such as:

```python
    @classmethod
    def should_lint(cls, view, settings, reason):
        # type: (sublime.View, LinterSettings, Reason) -> bool
        """Decide whether the linter can run at this point in time."""
        only_lint_saved_files = settings["only_lint_saved_files"]
        if only_lint_saved_files and (
            view.is_dirty() or not view.file_name()
        ):
            return False

        return super().should_lint(cls, view, settings, reason)

    def cmd(self):
        only_lint_saved_files = self.settings["only_lint_saved_files"]
        target = '$file_on_disk' if only_lint_saved_files else '$temp_file'
        return ('phpmd', target, 'text')
```

braver replied:
Fair points. Not sure there is a black-and-white good answer here. But always good to discuss.

> the user sets "manual", hits "lint this view" on dirty buffers but it actually still lints the file on disk

This is how real-file-only linters always work right?

>  All this even when the users doesn't use any baseline stuff.

Presumably most users will never change the lint mode anyway, unless (in this case) consciously doing so to support baseline stuff. That's the assumption that allows us to change the implementation based on the lint mode setting in the first place. Alternatives might have fewer assumptions, but come with their own downsides.

> The user could now roll eyes and fix it by saving and linting again.

This also holds the assumption that that's not disorienting, frustrating or confusing. I work on files that have hundreds of violations without the baseline. It would be very weird (and feel broken) if in one mode it reports 150 errors and in another it reports 0. I would argue it's easier to understand that you need to save the file for the linter to see your changes. That's how the entire load_save and save modes operate anyway: the error markers are potentially outdated until you save.

kaste replied:
> This is how real-file-only linters always work right?

No, real-file linters only run on saved files.  If you hit "lint this view" they (silently) skip. And they must do so because when we run  `('phpmd', '${file_on_disk}', 'text')` on a dirty buffer, we lint the file on disk and would highlight the wrong regions because the view is dirty,  That is really a bug imo. 

> Presumably most users will never change the lint mode anyway

Users already run in "manual" mode and we can't change the semantics of it, that's the problem.  It would be a breaking change, and we basically loose the "lint this view" at your will functionality.  Which is not while-you-type but whenever-you-want.  

Eye-rolling means an annoyance with a work around.  For all linters (at least typically) running on stdin or a temp_file when dirty or file_on_disk when just saved should yield the exact same output.   So here it does not, and if that's too confusing/eye-rolling we maybe should add a switch "only run on saved files".

kaste replied:
Dang, so it essentially comes down to if we forbid the manual "lint this view" *trigger*.  Because the same is true for load and save mode, we had to skip for a dirty buffers to not get garbage results.  Basically like this:

```python
   def cmd(self):
        if self.view.is_dirty() or not self.view.file_name():
            if self.settings.get("lint_mode") in ("load_save", "save"):
                window = self.view.window()
                if window:
                    window.status_message("please save first")
                raise TransientError("Abort.  Buffer is dirty.")
            return ('phpmd', '${temp_file}', 'text')
            
        self.tempfile_suffix = '-'
        return ('phpmd', '${file_on_disk}', 'text')
```

And I think this code makes it clear why this is fishy, subtle.  Why don't we use `$temp_file` and have to save first?  What has that to do with the "mode"?   And then you replace the conditional, like so

```python
   def cmd(self):
        if self.view.is_dirty() or not self.view.file_name():
            if self.settings["only_lint_saved_files"]:
                window = self.view.window()
                if window:
                    window.status_message("please save first")
                raise TransientError("Abort.  Buffer is dirty.")
            return ('phpmd', '${temp_file}', 'text')
            
        self.tempfile_suffix = '-'
        return ('phpmd', '${file_on_disk}', 'text')
```

and then it is at least clear that this is an arbitrary decision, a user choice.  

Does that makes sense?

braver replied:
> No, real-file linters only run on saved files. If you hit "lint this view" they (silently) skip. 

Ah, right, forgot about that. That changes things..

I like your suggestions, I’ll play with that a little bit tomorrow, see what that feels 
Ieke in practice.

braver replied:
Alright, I pushed something that works. 

I like using the "mode" as the behaviour switch here. A different new setting, like "only_lint_saved_files" would interact weirdly with that mode setting: if you set the mode to "background" and "only_lint_saved_files" to true, what does that even mean? 

The only subtle side effect of using the mode as I'm doing here, is that you no longer get to use the manual trigger on dirty files. Edge cases that we can message about. Setting to "save" or "load_save" mode already implies that you intend to run on real files anyway.

Eh, all options have downsides, we can bike-shed all day here 🤷🏻

kaste replied:
That's not bike-shedding, you do a breaking a change here and restrict the possible usages, for example that you want to lint automatically on save but also at your-will, per shortcut.  

"only_lint_saved_files" is maybe misleading.  Technically you switch the operation mode of the linter.  We have stdin, temp file and real file modes.   And these are not switchable by the user as they would typically involve also changing the `cmd`, but here you want to opt into real-file mode.  Is "switch_to_real_file_mode: true" better?  I don't think so because why should we introduce the modes here in such a way?  "support_baseline_files: true" says what the switch is supposed to work-around but hides the side effect of loosing the while-you-type feature, hence comes as a surprise. 

> if you set the mode to "background" and "only_lint_saved_files" to true, what does that even mean?

The default for all linters is "background" and the real-file linters "downgrade" automatically.  That is maybe confusing but also already a present conception.

braver replied:
Alright, color me convinced, you make good points all around. Like I said, I'm rusty on the subject and don't have all the modes and settings clear in my head. So, thanks 👍🏻 

I would say the new setting should have a name that closely represents _what_ it does, not _why_ (there can be many more, sometimes subtle "why's", there is only one "what"). And the "what" also indicates the side effect of no longer running on dirty buffers as you type. We can explain _why_ you'd want to use that setting in the README.

I think `"real_file_mode": true/false` is clear and simple.

kaste replied:
`real_file_mode` sounds really preferable.

It just leads to readable code, like I have a snippet some comments above, that would read now::

```
    def cmd(self):
        target = '$file_on_disk' if self.settings["real_file_mode"] else '$temp_file'
        return ('phpmd', target, 'text')

```

How can this be better, really?

braver replied:
So, pretty 🥹 Python can be really poetic ❤️

braver commented:
Finally got around to testing and tuning this. So, tldr;:

- added `real_file_mode` setting
- implemented that in `cmd()` and `should_lint()`
- updated readme around the new setting and support for ruleset and baseline files

Works on my machine 🤞🏻

kaste commented:
LGTM
